### PR TITLE
[public-api] IDEWorkspace service

### PR DESCRIPTION
**Pull request by jeanp413**
## Description
Follow up https://github.com/gitpod-io/gitpod/pull/14496#pullrequestreview-1177431495

## Release Notes

```release-note
NONE
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

File: components/public-api/gitpod/experimental/v1/ide_workspace.proto
Comment (akosyakov) [inline components/public-api/gitpod/experimental/v1/ide_workspace.proto:16]: `workspace_instance_id` or `instance_id`?

**Review by akosyakov — APPROVED** [review]
/hold

not sure about parameter names, i.e. workspace instance is not a workspace itself

File: components/public-api/gitpod/experimental/v1/ide_workspace.proto
Comment (easyCZ) [inline components/public-api/gitpod/experimental/v1/ide_workspace.proto:12]: `Close` May be sufficient, without leaking implementation details (it being a signal)

**Review by easyCZ — APPROVED** [review]
LGTM, modulo Q about Clone name and clarification on workspace argument name.

**Review by ArthurSens — APPROVED** [review]

File: components/public-api/gitpod/experimental/v1/ide_workspace.proto
Comment (csweichel) [inline components/public-api/gitpod/experimental/v1/ide_workspace.proto:9]: ws-manager calls this operation 'MarkActive'. What if we adopted the same terminology here?

File: components/public-api/gitpod/experimental/v1/ide_workspace.proto
Comment (csweichel) [inline components/public-api/gitpod/experimental/v1/ide_workspace.proto:12]: As a user of the API, how do I understand the difference between StopWorkspace and SendClose?

Could we fold this operation into SendHeartbeat?

File: components/public-api/gitpod/experimental/v1/ide_workspace.proto
Comment (csweichel) [inline components/public-api/gitpod/experimental/v1/ide_workspace.proto:16]: As far as I understand the public API does not know/talk about workspace instances. Ie 'workspace_id' would be the appropriate choice. 

Cc @easyCZ

Comment (easyCZ) [reply components/public-api/gitpod/experimental/v1/ide_workspace.proto:16]: For the `experimental` package of the Public API, this is less of a concern and I'm happy for us to also expose workspace instance.

Yes, ideally we'd not really talk about Workspace Instances in the Public API but in a way, that's not fully possible with existing WS APIs on Server. For now, what we do is we do expose the Workspace Instance ID but only in the Status response to a Workspace. From that perspective, it's in a way a secondary domain object, rather than being first class and customers having to understand them.

For the IDEWorkspaceService, which team IDE mostly intends to consume themselves at least for now I'm happy for us to continue iterating to find the right flow and get validation. If we did move this API to the `stable` package, we'd need to go through a review and think hard on how we name and structure the API, including exactly which fields do expose.

Comment (easyCZ) [reply components/public-api/gitpod/experimental/v1/ide_workspace.proto:16]: One way to avoid the `instance_id` is to send a `workspace_id` and the WorkspaceIDEService performs a `WorkspaceService.GetWorkspace(workspace_id)` call to retrieve the `instance_id`.

File: components/public-api/gitpod/experimental/v1/ide_workspace.proto
Comment (akosyakov) [reply components/public-api/gitpod/experimental/v1/ide_workspace.proto:12]: Maybe we rename a service to `IDEClientService` to clarify that it is related to a lifecycle of IDE client.

An idea was to separate functions because they are bound to different client lifecycle hooks. I guess initial design was rather about convenience of the implementor?

File: components/public-api/gitpod/experimental/v1/ide_workspace.proto
Comment (akosyakov) [reply components/public-api/gitpod/experimental/v1/ide_workspace.proto:16]: ok, thanks, it is good to realize it, maybe it is worth to document such key assumptions about API for future design

I guess we can go with @easyCZ suggestion now? We could also improve on it and do a check whether workspace running or not to prevent false positive heartbeats. Right now we have such logic both in VS Code browser and desktop to prevent. WDYT?

**Comment by akosyakov** [comment]
I like discussions on the PR, but I wonder how we can work more efficient. Do we need a short sync about it to align on key assumptions about new service? @csweichel @jeanp413 @easyCZ

@jeanp413 is there a PR for VS Code Desktop showing usages of these new APIs?

File: components/public-api/gitpod/experimental/v1/ide_workspace.proto
Comment (easyCZ) [reply components/public-api/gitpod/experimental/v1/ide_workspace.proto:16]: I'm happy with it as is. We'd review it before going into "stable" anyway. Moving forward on this and getting validation is as important as trying to get it right (but ultimately we won't until we actually use it)

**Pull request merged**
